pkg/api/featuremgmt.go

```diff
@@ -1,30 +1,95 @@
 package api
 
 import (
+	"fmt"
 	"net/http"
 
 	"github.com/grafana/grafana/pkg/api/response"
 	contextmodel "github.com/grafana/grafana/pkg/services/contexthandler/model"
+	"github.com/grafana/grafana/pkg/services/featuremgmt"
+	"github.com/grafana/grafana/pkg/setting"
+	"github.com/grafana/grafana/pkg/web"
 )
 
 func (hs *HTTPServer) GetFeatureToggles(ctx *contextmodel.ReqContext) response.Response {
-	featureMgmtCfg := hs.Cfg.FeatureManagement
-
-	features := hs.Features.GetFlags()
+	cfg := hs.Cfg.FeatureManagement
 	enabledFeatures := hs.Features.GetEnabled(ctx.Req.Context())
 
-	for i := 0; i < len(features); {
-		ft := features[i]
-		if _, ok := featureMgmtCfg.HiddenToggles[ft.Name]; ok {
-			features = append(features[:i], features[i+1:]...) // remove feature
+	// object being returned
+	dtos := make([]featuremgmt.FeatureToggleDTO, 0)
+
+	// loop through features an add features that should be visible to dtos
+	for _, ft := range hs.Features.GetFlags() {
+		if isFeatureHidden(ft, cfg.HiddenToggles) {
 			continue
 		}
-		if _, ok := featureMgmtCfg.ReadOnlyToggles[ft.Name]; ok {
-			features[i].ReadOnly = true
+		dto := featuremgmt.FeatureToggleDTO{
+			Name:        ft.Name,
+			Description: ft.Description,
+			Enabled:     enabledFeatures[ft.Name],
+			ReadOnly:    !isFeatureWriteable(ft, cfg.ReadOnlyToggles) || !isFeatureEditingAllowed(*hs.Cfg),
+		}
+
+		dtos = append(dtos, dto)
+	}
+
+	return response.JSON(http.StatusOK, dtos)
+}
+
+func (hs *HTTPServer) UpdateFeatureToggle(ctx *contextmodel.ReqContext) response.Response {
+	featureMgmtCfg := hs.Cfg.FeatureManagement
+	if !featureMgmtCfg.AllowEditing {
+		return response.Error(http.StatusForbidden, "feature toggles are read-only", fmt.Errorf("feature toggles are configured to be read-only"))
+	}
+
+	if featureMgmtCfg.UpdateControllerUrl == "" {
+		return response.Error(http.StatusInternalServerError, "feature toggles service is misconfigured", fmt.Errorf("[feature_management]update_controller_url is not set"))
+	}
+
+	cmd := featuremgmt.UpdateFeatureTogglesCommand{}
+	if err := web.Bind(ctx.Req, &cmd); err != nil {
+		return response.Error(http.StatusBadRequest, "bad request data", err)
+	}
+
+	for _, t := range cmd.FeatureToggles {
+		// make sure flag exists, and only continue if flag is writeable
+		if f, ok := hs.Features.LookupFlag(t.Name); ok && isFeatureWriteable(f, hs.Cfg.FeatureManagement.ReadOnlyToggles) {
+			hs.log.Info("UpdateFeatureToggle: updating toggle", "toggle_name", t.Name, "enabled", t.Enabled, "username", ctx.SignedInUser.Login)
+			// TODO build payload
```

Comment:
how will this work in OSS? is this essentially a cloud-only feature?

Reply:
Hey @ArturWierzbicki, our goal is to make this as widely adoptable as we can. At the very least, it needs to be useable by AWS and Azure managed Grafana instances, but we would like for it to be available for other Grafana operators as well. For what it's worth, we haven't done the phase 2 design yet, so the stuff we have in here for sending the request to a controller URL is still not definite.

One thing to consider is that someone running this in a small-scale environment on-prem probably won't have as much use for this, as they can just edit the ini file and restart. The target audience is more for larger orgs running Grafana as a service.